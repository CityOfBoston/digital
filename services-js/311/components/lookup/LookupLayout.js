// @flow

import React from 'react';
import { css } from 'glamor';
import type { Context } from 'next';
import Head from 'next/head';
import Router from 'next/router';

import type { RequestAdditions } from '../../server/next-handlers';

import Nav from '../common/Nav';
import CaseView from './CaseView';
import SearchForm from './SearchForm';

import type { Request } from '../../data/types';
import makeLoopbackGraphql from '../../data/dao/loopback-graphql';
import loadRequest from '../../data/dao/load-request';

const CONTAINER_STYLE = css({
  display: 'flex',
  flexDirection: 'column',
  minHeight: '100vh',
});

type SearchData = {
  view: 'search',
}

type CaseData = {
  view: 'case',
  query: string,
  request: ?Request,
}

export type InitialProps = {
  data: SearchData | CaseData,
};

export default class LookupLayout extends React.Component {
  static async getInitialProps({ query, req, res }: Context<RequestAdditions>): Promise<InitialProps> {
    let data;

    if (query.q) {
      const loopbackGraphql = makeLoopbackGraphql(req);
      const request = await loadRequest(loopbackGraphql, query.q);

      if (res && !request) {
        res.statusCode = 404;
      }

      data = {
        view: 'case',
        query: query.q,
        request,
      };
    } else {
      data = { view: 'search' };
    }

    return { data };
  }

  search = async (query: string) => {
    await Router.push(`/lookup?q=${query}`);
    window.scrollTo(0, 0);
  }

  render() {
    return (
      <div className={CONTAINER_STYLE}>
        <Head>
          <title>BOS:311 — {this.renderTitle()}</title>
        </Head>

        <Nav activeSection="lookup" />

        {this.renderContent()}
      </div>
    );
  }

  renderTitle() {
    const { data } = this.props;
    switch (data.view) {
      case 'search': return 'Case Lookup';
      case 'case': {
        const { request } = data;
        if (request) {
          return `Case #${request.id}`;
        } else {
          return 'Case not found';
        }
      }
      default: return '';
    }
  }

  renderContent() {
    const { data } = this.props;

    switch (data.view) {
      case 'search':
        return (
          <SearchForm searchFunc={this.search} />
        );
      case 'case': {
        const { query, request } = data;
        return (
          <CaseView query={query} request={request} />
        );
      }
      default:
        return null;
    }
  }
}
