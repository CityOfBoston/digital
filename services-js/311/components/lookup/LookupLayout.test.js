// @flow
/* eslint no-fallthrough: 0 */

import React from 'react';
import renderer from 'react-test-renderer';
import Router from 'next/router';

import type { Request } from '../../data/types';
import { makeServerContext } from '../../lib/test/make-context';

import LookupLayout from './LookupLayout';

jest.mock('next/router');
jest.mock('../../data/dao/load-request');

const loadRequest: JestMockFn = (require('../../data/dao/load-request'): any).default;

export const MOCK_REQUEST: Request = {
  id: '17-000000001',
  service: {
    name: 'Cosmic Intervention',
  },
  description: 'I think that Thanos is here',
  status: 'open',
  address: 'City Hall Plaza, Boston, MA 02131',
  requestedAtString: 'March 7, 2017, 12:59 PM',
  updatedAtString: 'April 8, 2017, 12:59 PM',
};

describe('search form', () => {
  let data;

  beforeEach(async () => {
    const ctx = makeServerContext('/lookup');
    data = (await LookupLayout.getInitialProps(ctx)).data;
  });

  test('getInitialProps', () => {
    expect(data.view).toEqual('search');
  });

  test('rendering', () => {
    const component = renderer.create(
      <LookupLayout data={data} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('case', () => {
  let data;

  beforeEach(async () => {
    loadRequest.mockReturnValue(Promise.resolve(MOCK_REQUEST));
    const ctx = makeServerContext('/search', { q: 'case-id' });
    data = (await LookupLayout.getInitialProps(ctx)).data;
  });

  test('getInitialProps', () => {
    switch (data.view) {
      case 'case':
        expect(data.query).toEqual('case-id');
      default:
        expect(data.view).toEqual('case');
    }
  });

  test('rendering', () => {
    const component = renderer.create(
      <LookupLayout data={data} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('case not found', () => {
  let ctx;
  let data;

  beforeEach(async () => {
    loadRequest.mockReturnValue(Promise.resolve(null));
    ctx = makeServerContext('/search', { q: 'not-a-real-id' });
    data = (await LookupLayout.getInitialProps(ctx)).data;
  });

  test('getInitialProps', () => {
    // flow check
    if (ctx.res) {
      expect(ctx.res.statusCode).toEqual(404);
    } else {
      expect(ctx.res).toBeDefined();
    }

    switch (data.view) {
      case 'case':
        expect(data.request).toBeNull();
      default:
        expect(data.view).toEqual('case');
    }
  });

  test('rendering', () => {
    const component = renderer.create(
      <LookupLayout data={data} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

test('search', () => {
  const lookupLayout = new LookupLayout({ data: { view: 'search' } });
  lookupLayout.search('query');
  expect(Router.push).toHaveBeenCalledWith('/search?q=query');
});
