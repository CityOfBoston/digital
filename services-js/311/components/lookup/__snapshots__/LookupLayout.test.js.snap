// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`case not found rendering 1`] = `
<div
  className="mn mn--full mn--nv-s"
>
  <nav
    className="nv-s nv-s--y nv-s--sticky"
  >
    <input
      aria-hidden={true}
      className="nv-s-tr"
      id="nv-s-tr"
      type="checkbox"
    />
    <ul
      className="nv-s-l"
    >
      <li
        className="nv-s-l-i"
      >
        <label
          className="nv-s-l-b"
          htmlFor="nv-s-tr"
          type="button"
        >
          Navigation
        </label>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/"
          onClick={[Function]}
        >
          Report a problem
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/services"
          onClick={[Function]}
        >
          See all services
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a nv-s-l-a--active"
          href="/lookup"
          onClick={[Function]}
        >
          Case look up
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/faq"
          onClick={[Function]}
        >
          FAQ
        </a>
      </li>
    </ul>
  </nav>
  <div
    className={
      Object {
        "data-css-d7jlfw": "",
      }
    }
  >
    <div
      className="sh sh--y"
    >
      <h2
        className="sh-title"
      >
        No results found
      </h2>
    </div>
    <div>
      <p
        className="t--info"
      >
        There is no case number matching “
        not-a-real-id
        ”.
      </p>
      <div
        className="stp css-38gbn5"
      >
        <span
          className="stp-number css-px1sz5"
        >
          1
        </span>
        <div
          className={null}
        >
          <form
            acceptCharset="UTF-8"
            action="/lookup"
            className="sf sf--y sf--md"
            method="get"
            onSubmit={[Function]}
          >
            <div
              className="sf-i"
            >
              <input
                className="sf-i-f"
                name="q"
                onInput={[Function]}
                placeholder="Search again…"
                type="text"
                value=""
              />
              <button
                className="sf-i-b"
              >
                Search
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        className="stp css-38gbn5"
      >
        <span
          className="stp-number css-lzdudf"
        >
          2
        </span>
        <span
          className="t--intro"
        >
          Our 311 operators are available 24/7 to help point you 
          <br />
          in the right direction. Call 
          <a
            href="tel:311"
          >
            311
          </a>
          , or 
          <a
            href="tel:+16176354500"
          >
            617-635-4500
          </a>
          .
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`case rendering 1`] = `
<div
  className="mn mn--full mn--nv-s"
>
  <nav
    className="nv-s nv-s--y nv-s--sticky"
  >
    <input
      aria-hidden={true}
      className="nv-s-tr"
      id="nv-s-tr"
      type="checkbox"
    />
    <ul
      className="nv-s-l"
    >
      <li
        className="nv-s-l-i"
      >
        <label
          className="nv-s-l-b"
          htmlFor="nv-s-tr"
          type="button"
        >
          Navigation
        </label>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/"
          onClick={[Function]}
        >
          Report a problem
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/services"
          onClick={[Function]}
        >
          See all services
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a nv-s-l-a--active"
          href="/lookup"
          onClick={[Function]}
        >
          Case look up
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/faq"
          onClick={[Function]}
        >
          FAQ
        </a>
      </li>
    </ul>
  </nav>
  <div
    className={
      Object {
        "data-css-d7jlfw": "",
      }
    }
  >
    <div
      className="sh sh--y"
    >
      <h2
        className="sh-title"
      >
        Case Search
      </h2>
    </div>
    <div
      className={
        Object {
          "data-css-1kcx6e8": "",
        }
      }
    >
      <h3
        className="t--intro m-v500"
      >
        Case Number: 
        <span
          className="t--number"
        >
          17-000000001
        </span>
      </h3>
      <hr
        className="hr hr--dash m-v300"
      />
      <div
        className="g"
      >
        <ul
          className="dl g--6"
        >
          <li
            className="dl-i"
          >
            <span
              className="dl-t"
            >
              Opened
            </span>
            <span
              className="dl-d dl-d--tt-n"
            >
              March 7, 2017, 12:59 PM
            </span>
          </li>
          <li
            className="dl-i"
          >
            <span
              className="dl-t"
            >
              Where
            </span>
            <span
              className="dl-d dl-d--tt-n"
            >
              City Hall Plaza, Boston, MA 02131
            </span>
          </li>
          <li
            className="dl-i"
          >
            <span
              className="dl-t"
            >
              Title
            </span>
            <span
              className="dl-d dl-d--tt-n"
            >
              Cosmic Intervention
            </span>
          </li>
          <li
            className="dl-i"
          >
            <span
              className="dl-t"
            >
              Status
            </span>
            <span
              className="dl-d dl-d--tt-n"
            >
              open
            </span>
          </li>
          <li
            className="dl-i"
          >
            <span
              className="dl-t"
            >
              Updated
            </span>
            <span
              className="dl-d dl-d--tt-n"
            >
              April 8, 2017, 12:59 PM
            </span>
          </li>
        </ul>
        <ul
          className="dl g--6"
        >
          <li
            className="dl-i"
          >
            <span
              className="dl-t"
            >
              Description
            </span>
            <span
              className="dl-d dl-d--tt-n"
            >
              I think that Thanos is here
            </span>
          </li>
        </ul>
      </div>
      <div
        className="g m-v500"
      >
        <a
          className="g--3 ta-c btn btn--y"
          href="/lookup"
          onClick={[Function]}
        >
          Back to Search
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`search form rendering 1`] = `
<div
  className="mn mn--full mn--nv-s"
>
  <nav
    className="nv-s nv-s--y nv-s--sticky"
  >
    <input
      aria-hidden={true}
      className="nv-s-tr"
      id="nv-s-tr"
      type="checkbox"
    />
    <ul
      className="nv-s-l"
    >
      <li
        className="nv-s-l-i"
      >
        <label
          className="nv-s-l-b"
          htmlFor="nv-s-tr"
          type="button"
        >
          Navigation
        </label>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/"
          onClick={[Function]}
        >
          Report a problem
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/services"
          onClick={[Function]}
        >
          See all services
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a nv-s-l-a--active"
          href="/lookup"
          onClick={[Function]}
        >
          Case look up
        </a>
      </li>
      <li
        className="nv-s-l-i"
      >
        <a
          className="nv-s-l-a "
          href="/faq"
          onClick={[Function]}
        >
          FAQ
        </a>
      </li>
    </ul>
  </nav>
  <div
    className={
      Object {
        "data-css-18erp1v": "",
      }
    }
  >
    <form
      acceptCharset="UTF-8"
      action="/lookup"
      className="sf sf--y "
      method="get"
      onSubmit={[Function]}
    >
      <div
        className="sf-i"
      >
        <input
          className="sf-i-f"
          name="q"
          onInput={[Function]}
          placeholder="Search case ID…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
        >
          Search
        </button>
      </div>
    </form>
  </div>
</div>
`;
